<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-файлов" data-toc-modified-id="Подготовка-файлов-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка файлов</a></span></li><li><span><a href="#Разведочный-анализ-данных" data-toc-modified-id="Разведочный-анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разведочный анализ данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-bomb" data-toc-modified-id="Таблица-bomb-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица bomb</a></span></li><li><span><a href="#Таблица-ranged_weapon" data-toc-modified-id="Таблица-ranged_weapon-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица ranged_weapon</a></span></li><li><span><a href="#Таблица-melee" data-toc-modified-id="Таблица-melee-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица melee</a></span></li><li><span><a href="#Таблица-creatures" data-toc-modified-id="Таблица-creatures-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица creatures</a></span></li><li><span><a href="#Соединяем-таблицы" data-toc-modified-id="Соединяем-таблицы-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Соединяем таблицы</a></span></li></ul></li><li><span><a href="#Построение-модели-предсказания" data-toc-modified-id="Построение-модели-предсказания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Построение модели предсказания</a></span><ul class="toc-item"><li><span><a href="#Подсчет-кол-во-патронов-для-полного-поражения-цели" data-toc-modified-id="Подсчет-кол-во-патронов-для-полного-поражения-цели-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Подсчет кол-во патронов для полного поражения цели</a></span></li><li><span><a href="#Построем-модель-предсказания" data-toc-modified-id="Построем-модель-предсказания-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Построем модель предсказания</a></span><ul class="toc-item"><li><span><a href="#Сделаем-Baseline" data-toc-modified-id="Сделаем-Baseline-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Сделаем Baseline</a></span></li><li><span><a href="#Подберем-лучшие-параметры-через-Hyperopt" data-toc-modified-id="Подберем-лучшие-параметры-через-Hyperopt-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Подберем лучшие параметры через Hyperopt</a></span></li><li><span><a href="#Построим-модель-с-найденными-параметрами" data-toc-modified-id="Построим-модель-с-найденными-параметрами-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Построим модель с найденными параметрами</a></span></li><li><span><a href="#Сделаем-кросс-валидацию-на-всех-данных" data-toc-modified-id="Сделаем-кросс-валидацию-на-всех-данных-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Сделаем кросс-валидацию на всех данных</a></span></li><li><span><a href="#Сделаем-финальное-предсказание" data-toc-modified-id="Сделаем-финальное-предсказание-3.2.5"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Сделаем финальное предсказание</a></span></li></ul></li></ul></li></ul></div>

In [363]:
# скачаем таблицы
!gdown 1j_cL4vDAKsNZn1Jo2wM4njbl5PiYwQdt
!gdown 1F7Jo77NQ9q2EqqTx978JNFNxFnFf9gUi
!gdown 12rV0tJ5MtxzU3v8Hs_rkOaXqz3gF7eDc
!gdown 1N3YTtwSRSXtor_x659OFlyg6wD88xZlp

Downloading...
From: https://drive.google.com/uc?id=1j_cL4vDAKsNZn1Jo2wM4njbl5PiYwQdt
To: C:\Users\lavru\juoyter_notebook\Weapon_Fallout4\bomb.csv

  0%|          | 0.00/1.75k [00:00<?, ?B/s]
100%|##########| 1.75k/1.75k [00:00<?, ?B/s]
Downloading...
From: https://drive.google.com/uc?id=1F7Jo77NQ9q2EqqTx978JNFNxFnFf9gUi
To: C:\Users\lavru\juoyter_notebook\Weapon_Fallout4\ranged_weapon.csv

  0%|          | 0.00/9.81k [00:00<?, ?B/s]
100%|##########| 9.81k/9.81k [00:00<00:00, 4.92MB/s]
Downloading...
From: https://drive.google.com/uc?id=12rV0tJ5MtxzU3v8Hs_rkOaXqz3gF7eDc
To: C:\Users\lavru\juoyter_notebook\Weapon_Fallout4\melee.csv

  0%|          | 0.00/3.47k [00:00<?, ?B/s]
100%|##########| 3.47k/3.47k [00:00<00:00, 3.49MB/s]
Downloading...
From: https://drive.google.com/uc?id=1N3YTtwSRSXtor_x659OFlyg6wD88xZlp
To: C:\Users\lavru\juoyter_notebook\Weapon_Fallout4\creatures.csv

  0%|          | 0.00/24.3k [00:00<?, ?B/s]
100%|##########| 24.3k/24.3k [00:00<00:00, 6.09MB/s]


# Подготовка файлов

In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
import warnings
warnings.filterwarnings('ignore')

Сслыки на файлы с таблицами:
+ https://drive.google.com/file/d/1j_cL4vDAKsNZn1Jo2wM4njbl5PiYwQdt/view?usp=share_link (bomb.csv)
+ https://drive.google.com/file/d/1F7Jo77NQ9q2EqqTx978JNFNxFnFf9gUi/view?usp=share_link (ranged_weapon.csv)
+ https://drive.google.com/file/d/12rV0tJ5MtxzU3v8Hs_rkOaXqz3gF7eDc/view?usp=share_link (melee.csv)
+ https://drive.google.com/file/d/1N3YTtwSRSXtor_x659OFlyg6wD88xZlp/view?usp=share_link (creatures.csv)

In [3]:
bomb = pd.read_csv("bomb.csv")
ranged_weapon = pd.read_csv("ranged_weapon.csv")
melee = pd.read_csv("melee.csv")
creatures = pd.read_csv("creatures.csv")

# Разведочный анализ данных

## Таблица bomb

In [4]:
bomb

Unnamed: 0                               Название         Тип Тип_Оружия  \
0            0           Бутылка с коктейлем Молотова  Взрывчатка    Граната   
1            1           Граната из бейсбольного мяча  Взрывчатка    Граната   
2            2                     Осколочная граната  Взрывчатка    Граната   
3            3                     Плазменная граната  Взрывчатка    Граната   
4            4                     Импульсная граната  Взрывчатка    Граната   
5            5        Институтская импульсная граната  Взрывчатка    Граната   
6            6                     Криогенная граната  Взрывчатка    Граната   
7            7                         Маяк Института  Взрывчатка    Граната   
8            8                           Молотов-кола  Взрывчатка    Граната   
9            9                        Осколочная мина  Взрывчатка       Мина   
10          10                        Плазменная мина  Взрывчатка       Мина   
11          11                            Крышко-мина  Взрывчатка       Мина   
12          12                               Криомина  Взрывчатка       Мина   
13          13                        Импульсная мина  Взрывчатка       Мина   
14          14                           Ядерная мина  Взрывчатка       Мина   
15          15                           Ядерный блок  Взрывчатка    Граната   
16          16  Разделяющаяся осколочная граната (nw)  Взрывчатка    Граната   
17          17        «Умная» осколочная граната (nw)  Взрывчатка    Граната   

                    Урон  Вес Цена  
0                     51  0,5   20  
1                    101    1   40  
2                    151  0,5   50  
3   150+150 (electrical)  0,5  135  
4       150 (electrical)  0,5  100  
5       150 (electrical)  0,5  100  
6            101 (frost)  0,5  125  
7                      1  NaN  NaN  
8    301+100 (radiation)  0,5  100  
9                    101  0,5   50  
10  150+150 (electrical)  0,5  100  
11                   301  0,5   75  
12           101 (frost)  0,5   50  
13      150 (electrical)  0,5  100  
14   301+100 (radiation)  0,5  100  
15                     -    4    -  
16                     1    1   75  
17                   182  0,6   60

В столбце "Урон" присутствуют дополнительные статы (frost, radiation, electrical). Создадим дополнительные столбцы с этими статами

In [5]:
bomb.isna().mean()

Unnamed: 0    0.000000
Название      0.000000
Тип           0.000000
Тип_Оружия    0.000000
Урон          0.000000
Вес           0.055556
Цена          0.055556
dtype: float64

Посмотрим на пропущенные данные

In [6]:
bomb[bomb['Цена'].isna()]

Unnamed: 0        Название         Тип Тип_Оружия Урон  Вес Цена
7           7  Маяк Института  Взрывчатка    Граната    1  NaN  NaN

Одно пропущенное значение для столбцов "Вес" и "Цена", посмотрим в [интернете](https://fallout.fandom.com/ru/wiki/%D0%9C%D0%B0%D1%8F%D0%BA_%D0%98%D0%BD%D1%81%D1%82%D0%B8%D1%82%D1%83%D1%82%D0%B0) и добавим отсутствующие значения для этого оружия

In [7]:
bomb['Вес'] = bomb['Вес'].fillna(0.5)

In [8]:
bomb['Цена'] = bomb['Цена'].fillna(0)

In [9]:
bomb.dtypes

Unnamed: 0     int64
Название      object
Тип           object
Тип_Оружия    object
Урон          object
Вес           object
Цена          object
dtype: object

Почти все колонки имеют тип данныхю "object", необходимо выделить дополнительные статы урона и привести нужные столбцы к типу данных "int"

In [10]:
bomb

Unnamed: 0                               Название         Тип Тип_Оружия  \
0            0           Бутылка с коктейлем Молотова  Взрывчатка    Граната   
1            1           Граната из бейсбольного мяча  Взрывчатка    Граната   
2            2                     Осколочная граната  Взрывчатка    Граната   
3            3                     Плазменная граната  Взрывчатка    Граната   
4            4                     Импульсная граната  Взрывчатка    Граната   
5            5        Институтская импульсная граната  Взрывчатка    Граната   
6            6                     Криогенная граната  Взрывчатка    Граната   
7            7                         Маяк Института  Взрывчатка    Граната   
8            8                           Молотов-кола  Взрывчатка    Граната   
9            9                        Осколочная мина  Взрывчатка       Мина   
10          10                        Плазменная мина  Взрывчатка       Мина   
11          11                            Крышко-мина  Взрывчатка       Мина   
12          12                               Криомина  Взрывчатка       Мина   
13          13                        Импульсная мина  Взрывчатка       Мина   
14          14                           Ядерная мина  Взрывчатка       Мина   
15          15                           Ядерный блок  Взрывчатка    Граната   
16          16  Разделяющаяся осколочная граната (nw)  Взрывчатка    Граната   
17          17        «Умная» осколочная граната (nw)  Взрывчатка    Граната   

                    Урон  Вес Цена  
0                     51  0,5   20  
1                    101    1   40  
2                    151  0,5   50  
3   150+150 (electrical)  0,5  135  
4       150 (electrical)  0,5  100  
5       150 (electrical)  0,5  100  
6            101 (frost)  0,5  125  
7                      1  0.5    0  
8    301+100 (radiation)  0,5  100  
9                    101  0,5   50  
10  150+150 (electrical)  0,5  100  
11                   301  0,5   75  
12           101 (frost)  0,5   50  
13      150 (electrical)  0,5  100  
14   301+100 (radiation)  0,5  100  
15                     -    4    -  
16                     1    1   75  
17                   182  0,6   60

Создадим датафрайм split_bomb для последующей работы 

In [11]:
"Функция проверяет есть ли вхождения из списка damage"
def allpresent(text, damage):
    for damag in damage:
        while text.find(damag) != -1:
            return True
    return False

In [12]:
"Функция фильтрует данные урона"
def damage_sorting(tabel):
    columns = ['damage','electrical','radiation','frost','fire','poison',"blood","explosion"]
    df = pd.DataFrame(columns = columns)
    damage = columns[1:]

    for column in columns:
        df[column] = tabel
        
        if column.lower() not in damage:
            df[column] = df[column].str.split('+').apply(lambda x: x[0] if (len(x) == 2) or (len(x) == 3) else x)
            df = df.applymap(lambda x: x[0] if isinstance(x, list) else x)
            df[column] = df[column].apply(lambda x: x.replace(f"{x}",'0') if allpresent(x, damage) else x) 
    
        else:
            df[column] = df[column].apply(lambda x: x.replace(f"{x}",'0') if column.lower() not in x else x) #
            df[column] = df[column].str.split('+').apply(lambda x: x[1] if (len(x) == 2) else x)
            df = df.applymap(lambda x: x[0] if isinstance(x, list) else x)
            df[column] = df[column].apply(lambda x: x.replace(f' ({column.lower()})', ''))
    return df

In [13]:
split_bomb = damage_sorting(bomb["Урон"])

In [14]:
split_bomb

damage electrical radiation frost fire poison blood explosion
0      51          0         0     0    0      0     0         0
1     101          0         0     0    0      0     0         0
2     151          0         0     0    0      0     0         0
3     150        150         0     0    0      0     0         0
4       0        150         0     0    0      0     0         0
5       0        150         0     0    0      0     0         0
6       0          0         0   101    0      0     0         0
7       1          0         0     0    0      0     0         0
8     301          0       100     0    0      0     0         0
9     101          0         0     0    0      0     0         0
10    150        150         0     0    0      0     0         0
11    301          0         0     0    0      0     0         0
12      0          0         0   101    0      0     0         0
13      0        150         0     0    0      0     0         0
14    301          0       100     0    0      0     0         0
15      -          0         0     0    0      0     0         0
16      1          0         0     0    0      0     0         0
17    182          0         0     0    0      0     0         0

Сделаем конкатенацию таблиц bomb и split_bomb

In [15]:
bomb = pd.concat([bomb,split_bomb],axis = 1)

In [16]:
bomb = bomb.drop(['Урон','Тип_Оружия'], axis = 1)

In [17]:
bomb.rename(columns = {'Тип':'Тип_Оружия'}, inplace = True )

In [18]:
bomb["Цена"] = np.where((bomb["Цена"] == '-'), 0, bomb["Цена"])

In [19]:
bomb["damage"] = np.where((bomb["damage"] == '-'), 0, bomb["damage"])

In [20]:
bomb['Вес'] = bomb['Вес'].apply(lambda x: str(x).replace(',','.'))

In [21]:
bomb[['Вес','Цена','damage','electrical','radiation','frost','fire','poison',"blood","explosion"]] = bomb[['Вес','Цена','damage','electrical','radiation','frost','fire','poison',"blood","explosion"]].astype(float)

In [22]:
bomb

Unnamed: 0                               Название  Тип_Оружия  Вес   Цена  \
0            0           Бутылка с коктейлем Молотова  Взрывчатка  0.5   20.0   
1            1           Граната из бейсбольного мяча  Взрывчатка  1.0   40.0   
2            2                     Осколочная граната  Взрывчатка  0.5   50.0   
3            3                     Плазменная граната  Взрывчатка  0.5  135.0   
4            4                     Импульсная граната  Взрывчатка  0.5  100.0   
5            5        Институтская импульсная граната  Взрывчатка  0.5  100.0   
6            6                     Криогенная граната  Взрывчатка  0.5  125.0   
7            7                         Маяк Института  Взрывчатка  0.5    0.0   
8            8                           Молотов-кола  Взрывчатка  0.5  100.0   
9            9                        Осколочная мина  Взрывчатка  0.5   50.0   
10          10                        Плазменная мина  Взрывчатка  0.5  100.0   
11          11                            Крышко-мина  Взрывчатка  0.5   75.0   
12          12                               Криомина  Взрывчатка  0.5   50.0   
13          13                        Импульсная мина  Взрывчатка  0.5  100.0   
14          14                           Ядерная мина  Взрывчатка  0.5  100.0   
15          15                           Ядерный блок  Взрывчатка  4.0    0.0   
16          16  Разделяющаяся осколочная граната (nw)  Взрывчатка  1.0   75.0   
17          17        «Умная» осколочная граната (nw)  Взрывчатка  0.6   60.0   

    damage  electrical  radiation  frost  fire  poison  blood  explosion  
0     51.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
1    101.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
2    151.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
3    150.0       150.0        0.0    0.0   0.0     0.0    0.0        0.0  
4      0.0       150.0        0.0    0.0   0.0     0.0    0.0        0.0  
5      0.0       150.0        0.0    0.0   0.0     0.0    0.0        0.0  
6      0.0         0.0        0.0  101.0   0.0     0.0    0.0        0.0  
7      1.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
8    301.0         0.0      100.0    0.0   0.0     0.0    0.0        0.0  
9    101.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
10   150.0       150.0        0.0    0.0   0.0     0.0    0.0        0.0  
11   301.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
12     0.0         0.0        0.0  101.0   0.0     0.0    0.0        0.0  
13     0.0       150.0        0.0    0.0   0.0     0.0    0.0        0.0  
14   301.0         0.0      100.0    0.0   0.0     0.0    0.0        0.0  
15     0.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
16     1.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0  
17   182.0         0.0        0.0    0.0   0.0     0.0    0.0        0.0

Удалим оружие "Ядерный блок" т.к. это оружие не наносит никакого вида урона

In [23]:
bomb = bomb[bomb['Название'] != 'Ядерный блок']

In [24]:
bomb.dtypes

Unnamed: 0      int64
Название       object
Тип_Оружия     object
Вес           float64
Цена          float64
damage        float64
electrical    float64
radiation     float64
frost         float64
fire          float64
poison        float64
blood         float64
explosion     float64
dtype: object

Привели в рабочий вид таблицу bomb

## Таблица ranged_weapon

In [25]:
ranged_weapon

Unnamed: 0                                Название Тип_Оружия  \
0            0                Гладкоствольный пистолет  Пистолеты   
1            1     Гладкоствольный с затвором пистолет  Пистолеты   
2            2  Гладкоствольный с патронником пистолет  Пистолеты   
3            3                          10-мм пистолет  Пистолеты   
4            4                  Друг обитателя Пустоши  Пистолеты   
..         ...                                     ...        ...   
81          81                               Криолятор    Тяжёлое   
82          82                              Гарпуномёт    Тяжёлое   
83          83                         Друг адмирала *    Тяжёлое   
84          84                    Гарпуномёт защитника    Тяжёлое   
85          85                 Последний оплот шкипера    Тяжёлое   

             Патроны        Урон Урон_Секунду  Дальность  Точность  \
0         Патрон .38          13          6.8       83.0      57.0   
1        Патрон .308          35         71.5       95.0      70.0   
2         Патрон .45          25         14.4       83.0      61.0   
3       10-мм патрон          18         82.8       83.0      60.0   
4       10-мм патрон    Случайно          NaN        NaN       NaN   
..               ...         ...          ...        ...       ...   
81  Криогенный заряд  20 (frost)           90       71.0      66.0   
82            Гарпун         150          NaN      143.0      63.0   
83            Гарпун         150          NaN      143.0      63.0   
84            Гарпун         150          NaN      143.0      63.0   
85            Гарпун         165          NaN      143.0      63.0   

   Магазин_Патронов   Вес    Цена  
0                12   2.3    20.0  
1                 6   4.0    36.0  
2                 6   4.2    25.0  
3                12   3.5    50.0  
4               NaN   NaN     NaN  
..              ...   ...     ...  
81               25  13.2   302.0  
82                1  16,3   205.0  
83                1  16,3  1005.0  
84                1  16,3  1005.0  
85                1  16,3  1005.0  

[86 rows x 11 columns]

In [26]:
len(ranged_weapon[ranged_weapon['Урон'] == "Случайно"])

17

В таблице часто встречается слово "Случайно", а именно 17 раз. Обратимся к сайту за разъяснения.
>Конкретный набор модификаций уникального оружия в Fallout 4 определяется случайным образом в момент его получения, или первого посещения локации, где оно хранится, или первого контакта с персонажем, который им владеет. Поэтому его внешний вид и характеристики могут отличаться у разных игроков.

Скорее всего характеристики этого случайного оружия не будут меньше, чем их оригинальный вариант. Получается они отличаются лишь названием и характеристикам от оригинального оружия, к тому же, тяжело сказать насколько конкретно будут различаться эти характеристики, поэтому просто удалим эти оружия  
Так же можно сказать и про другие предметы, где написано "Аналогично стандартному ПП" и т.д., они полностью повторяют характеристики, поэтому удалим и их

In [27]:
regex = "[а-я]"
pattern = re.compile(regex)

In [28]:
ranged_weapon = ranged_weapon[ranged_weapon['Урон'].apply(lambda x: pattern.search(x) is None)]

In [29]:
ranged_weapon.head(20)

Unnamed: 0                                Название Тип_Оружия  \
0            0                Гладкоствольный пистолет  Пистолеты   
1            1     Гладкоствольный с затвором пистолет  Пистолеты   
2            2  Гладкоствольный с патронником пистолет  Пистолеты   
3            3                          10-мм пистолет  Пистолеты   
5            5                               Спаситель  Пистолеты   
6            6                            .44 пистолет  Пистолеты   
10          10                    Ковбойский револьвер  Пистолеты   
11          11                   Институтский пистолет  Пистолеты   
12          12                       Лазерный пистолет  Пистолеты   
15          15                     Плазменный пистолет  Пистолеты   
16          16                             Гамма-пушка  Пистолеты   
17          17                          Дзета-пистолет  Пистолеты   
18          18                        Пистолет Лоренцо  Пистолеты   
19          19                      Внеземной пистолет  Пистолеты   
20          20                Бластер Чужих — Хабологи  Пистолеты   
21          21              Головной лазер штурмотрона  Пистолеты   
22          22                 Гладкоствольный карабин   Карабины   
23          23      Гладкоствольный c затвором карабин   Карабины   
24          24   Гладкоствольный с патронником карабин   Карабины   
25          25                   Двуствольный дробовик   Карабины   

                         Патроны  \
0                     Патрон .38   
1                    Патрон .308   
2                     Патрон .45   
3                   10-мм патрон   
5                   10-мм патрон   
6                     Патрон .44   
10                    Патрон .44   
11               Ядерная батарея   
12               Ядерная батарея   
15                  Заряд плазмы   
16        Патрон для гамма-пушки   
17        Патрон для гамма-пушки   
18        Патрон для гамма-пушки   
19  Боеприпас для бластера Чужих   
20  Боеприпас для бластера Чужих   
21               Ядерная батарея   
22                    Патрон .38   
23                   Патрон .308   
24                    Патрон .45   
25              Патрон дробовика   

                                            Урон  \
0                                             13   
1                                             35   
2                                             25   
3                                             18   
5                                             27   
6                                             48   
10                                            60   
11                               15 (electrical)   
12                               24 (electrical)   
15                            25+25 (electrical)   
16   110 (radiation) +23 (radiation) (explosion)   
17  10 (radiation) +11.5 (radiation) (explosion)   
18                25 (explosion) +10 (radiation)   
19                                            55   
20                                            50   
21                                            35   
22                                            13   
23                                            35   
24                                            25   
25                                          47,5   

                                     Урон_Секунду  Дальность  Точность  \
0                                             6.8       83.0      57.0   
1                                            71.5       95.0      70.0   
2                                            14.4       83.0      61.0   
3                                            82.8       83.0      60.0   
5                                             165       53.0      67.0   
6                                            28.8      119.0      66.0   
10                                            NaN      119.0      66.0   
11                                99 (electrical)       71.0      70.0   
12                               120 (

Посмотрим на пропущенные значения

In [30]:
ranged_weapon.isna().mean()

Unnamed: 0          0.000000
Название            0.000000
Тип_Оружия          0.000000
Патроны             0.000000
Урон                0.000000
Урон_Секунду        0.307692
Дальность           0.015385
Точность            0.015385
Магазин_Патронов    0.015385
Вес                 0.000000
Цена                0.015385
dtype: float64

**Пропуски в столбце "Урон_Секунду"**

In [31]:
ranged_weapon[ranged_weapon['Урон_Секунду'].isna()].head(20)

Unnamed: 0                                 Название Тип_Оружия  \
10          10                     Ковбойский револьвер  Пистолеты   
32          32  Специальное предложение от Техника Тома   Карабины   
33          33              Карабин с рычажным затвором   Карабины   
34          34                              Старый друг   Карабины   
35          35                         Счастливчик Эдди   Карабины   
40          40                         Радиевый карабин   Карабины   
41          41                     Переделанный аппарат   Карабины   
42          42             Радиевый карабин «Килотонна»   Карабины   
44          44                      Самодельный карабин   Карабины   
45          45                              Брызгострел   Карабины   
46          46         Человек, который решает проблемы   Карабины   
56          56                         Эксперимент 18-А   Карабины   
61          61                            Карабин Теслы   Карабины   
72          72                         Смерть с небес *    Тяжёлое   
75          75                                 Страйкер    Тяжёлое   
76          76                                 Ядер-Мёт    Тяжёлое   
82          82                               Гарпуномёт    Тяжёлое   
83          83                          Друг адмирала *    Тяжёлое   
84          84                     Гарпуномёт защитника    Тяжёлое   
85          85                  Последний оплот шкипера    Тяжёлое   

                              Патроны                             Урон  \
10                         Патрон .44                               60   
32                        Патрон .308                               55   
33                       Патрон 45-70                               36   
34                       Патрон 45-70                               70   
35                       Патрон 45-70                               74   
40                         Патрон .45                28+50 (radiation)   
41                         Патрон .45                56+56 (radiation)   
42                         Патрон .45  56+56(radiation) +15(explosion)   
44                        Патрон 7,62                               33   
45                        Патрон 7,62                               41   
46                        Патрон 7,62                               21   
56                       Заряд плазмы               24+24 (electrical)   
61                    Ядерная батарея                               38   
72                             Ракета                              151   
75  Модифицированный шар для боулинга                              100   
76                         Ядер-Заряд                              833   
82                             Гарпун                              150   
83                             Гарпун                              150   
84                             Гарпун                              150   
85                             Гарпун                              165   

   Урон_Секунду  Дальность  Точность Магазин_Патронов   Вес    Цена  
10          NaN      119.0      66.0                6   5.2    99.0  
32          NaN      185.0     112.0                5  16.7   709.0  
33          NaN      122.0      72.0                5     9   201.0  
34          NaN      122.0      61.0                5  12,7   801.0  
35          NaN      205.0      74.0                5  10,4   201.0  
40          NaN      119.0      69.0              NaN  11,1   110.0  
41          NaN      119.0      69.0               20  11,1   430.0  
42          NaN      119.0      69.0               20  11,1     NaN  
44          NaN      119.0      72.0               10   128   136.0  
45          NaN       89.0     108.0               75  23,6  6855.0  
46          NaN      191.0      74.0               30  17,8   612.0  
56          NaN      203.0     176.0                9   7,6  1919.0  
61          NaN       70.0       NaN               15   8,1    90.0 

В столбце "УронСекунду" возможные пропуски из-за того, что оружие имеет один выстрел. Согласно статье [fallout.fandom.com](https://fallout.fandom.com/ru/wiki/%D0%A3%D1%80%D0%BE%D0%BD_%D0%B2_%D1%81%D0%B5%D0%BA%D1%83%D0%BD%D0%B4%D1%83) 
>УВС для некоторых видов оружия (например, охотничья винтовка) гораздо ниже, чем УРО/ПАТР. Это происходит по причине того, что для некоторых видов оружия требуется больше 1 секунды для выстрела одного патрона и подачи другого.

Получается для этого оружия выстрел происходит слишком долго (больше секунды). Заполним данные ппропуски 0, т.к. за 1 секунду оружие не успевает нанести урон

In [32]:
ranged_weapon['Урон_Секунду'] = ranged_weapon['Урон_Секунду'].fillna(0)

**Пропуски в столбце "Дальность"**

In [33]:
ranged_weapon[ranged_weapon['Дальность'].isna()]

Unnamed: 0       Название Тип_Оружия       Патроны             Урон  \
67          67  Гатлинг-лазер    Тяжёлое  Ядерный блок  28 (electrical)   

        Урон_Секунду  Дальность  Точность Магазин_Патронов   Вес   Цена  
67  272 (electrical)        NaN      48.0              500  19.3  332.0

Здесь пропущенная дальность, можем заменить похожим оружием

In [34]:
ranged_weapon[ranged_weapon['Патроны'] == 'Ядерный блок']

Unnamed: 0       Название Тип_Оружия       Патроны             Урон  \
67          67  Гатлинг-лазер    Тяжёлое  Ядерный блок  28 (electrical)   
68          68  Последний суд    Тяжёлое  Ядерный блок               15   
69          69        Этернус    Тяжёлое  Ядерный блок               79   

        Урон_Секунду  Дальность  Точность Магазин_Патронов   Вес    Цена  
67  272 (electrical)        NaN      48.0              500  19.3   332.0  
68               476      203.0      48.0              500  19.3  3804.0  
69             355.5      251.0      63.0              500  39.3  4799.0

Заменим средним значением поле "дальность" по одному типу оружия

In [35]:
mn = np.mean(ranged_weapon[ranged_weapon['Патроны'] == 'Ядерный блок']['Дальность'].iloc[1:])

In [36]:
ranged_weapon['Дальность'] = ranged_weapon['Дальность'].fillna(mn)

**Пропуски в столбце "Точность"**

In [37]:
ranged_weapon[ranged_weapon['Точность'].isna()]

Unnamed: 0       Название Тип_Оружия          Патроны Урон Урон_Секунду  \
61          61  Карабин Теслы   Карабины  Ядерная батарея   38            0   

    Дальность  Точность Магазин_Патронов  Вес  Цена  
61       70.0       NaN               15  8,1  90.0

Заменим средним значением по такому же типу оружия и типу патрона

In [38]:
mn = np.mean(ranged_weapon[(ranged_weapon['Патроны'] == 'Ядерная батарея') 
              & (ranged_weapon['Тип_Оружия'] == 'Карабины')]['Точность'][:-1])

In [39]:
ranged_weapon['Точность'] = ranged_weapon['Точность'].fillna(mn)

**Пропуски в столбце "Магазин_Патронов"**

In [40]:
ranged_weapon[ranged_weapon['Магазин_Патронов'].isna()]

Unnamed: 0          Название Тип_Оружия     Патроны               Урон  \
40          40  Радиевый карабин   Карабины  Патрон .45  28+50 (radiation)   

   Урон_Секунду  Дальность  Точность Магазин_Патронов   Вес   Цена  
40            0      119.0      69.0              NaN  11,1  110.0

Заменим медианным значением по такому же типу оружия и типу патрона

In [41]:
mn = ranged_weapon[(ranged_weapon['Патроны'] == 'Патрон .45')
              & (ranged_weapon['Тип_Оружия'] == 'Карабины')]['Магазин_Патронов'].sort_values()

In [42]:
mn = mn[:-1].median()

In [43]:
ranged_weapon['Магазин_Патронов'] = ranged_weapon['Магазин_Патронов'].fillna(mn)

**Пропуски в столбце "Цена"**

In [44]:
ranged_weapon[ranged_weapon['Цена'].isna()]

Unnamed: 0                      Название Тип_Оружия     Патроны  \
42          42  Радиевый карабин «Килотонна»   Карабины  Патрон .45   

                               Урон Урон_Секунду  Дальность  Точность  \
42  56+56(radiation) +15(explosion)            0      119.0      69.0   

   Магазин_Патронов   Вес  Цена  
42               20  11,1   NaN

Заменим средним значением по такому же типу оружия и типу патрона

In [45]:
mn = np.mean(ranged_weapon[(ranged_weapon['Патроны'] == 'Патрон .45') 
              & (ranged_weapon['Тип_Оружия'] == 'Карабины')]['Цена'][:-1])

In [46]:
ranged_weapon['Цена'] = ranged_weapon['Цена'].fillna(mn)

In [47]:
ranged_weapon.isna().mean()

Unnamed: 0          0.0
Название            0.0
Тип_Оружия          0.0
Патроны             0.0
Урон                0.0
Урон_Секунду        0.0
Дальность           0.0
Точность            0.0
Магазин_Патронов    0.0
Вес                 0.0
Цена                0.0
dtype: float64

Все пропуски заполнены

In [48]:
ranged_weapon.dtypes

Unnamed: 0            int64
Название             object
Тип_Оружия           object
Патроны              object
Урон                 object
Урон_Секунду         object
Дальность           float64
Точность            float64
Магазин_Патронов     object
Вес                  object
Цена                float64
dtype: object

Некоторые столбцы имеют тип object другие float64, приведем к правильному виду в конце

В данной таблице присутствует как базовый урон, так и урон в секунду, изменим эти две колонки. Для этого создадим 2 датафрейма split_bomb_damage и split_bomb_damage_sec

In [49]:
split_bomb_damage = damage_sorting(ranged_weapon["Урон"])

In [50]:
split_bomb_damage.head()

damage electrical radiation frost fire poison blood explosion
0     13          0         0     0    0      0     0         0
1     35          0         0     0    0      0     0         0
2     25          0         0     0    0      0     0         0
3     18          0         0     0    0      0     0         0
5     27          0         0     0    0      0     0         0

Для некоторых оружий присутствует как explosion, так и radiation одновременно. Функция правильно их распределила, однако оставила надписи. Уберем их вручную

In [51]:
split_bomb_damage['radiation'] = split_bomb_damage['radiation'].apply(lambda x: x.replace(" (explosion)", ''))

In [52]:
split_bomb_damage['explosion'] = split_bomb_damage['explosion'].apply(lambda x: x.replace(" (radiation)", ''))

In [53]:
split_bomb_damage.head(20)

damage electrical radiation frost fire poison blood explosion
0      13          0         0     0    0      0     0         0
1      35          0         0     0    0      0     0         0
2      25          0         0     0    0      0     0         0
3      18          0         0     0    0      0     0         0
5      27          0         0     0    0      0     0         0
6      48          0         0     0    0      0     0         0
10     60          0         0     0    0      0     0         0
11      0         15         0     0    0      0     0         0
12      0         24         0     0    0      0     0         0
15     25         25         0     0    0      0     0         0
16      0          0        23     0    0      0     0        23
17      0          0      11.5     0    0      0     0      11.5
18      0          0        10     0    0      0     0        10
19     55          0         0     0    0      0     0         0
20     50          0         0     0    0      0     0         0
21     35          0         0     0    0      0     0         0
22     13          0         0     0    0      0     0         0
23     35          0         0     0    0      0     0         0
24     25          0         0     0    0      0     0         0
25   47,5          0         0     0    0      0     0         0

Проделаем все тоже самое с уроном в секунду

In [54]:
ranged_weapon['Урон_Секунду'] = ranged_weapon['Урон_Секунду'].apply(lambda x: str(x).replace(",", '.'))

In [55]:
split_bomb_damage_sec = damage_sorting(ranged_weapon["Урон_Секунду"])

In [56]:
split_bomb_damage_sec.head(20)

damage electrical          radiation frost fire poison blood  \
0      6.8          0                  0     0    0      0     0   
1     71.5          0                  0     0    0      0     0   
2     14.4          0                  0     0    0      0     0   
3     82.8          0                  0     0    0      0     0   
5      165          0                  0     0    0      0     0   
6     28.8          0                  0     0    0      0     0   
10       0          0                  0     0    0      0     0   
11       0         99                  0     0    0      0     0   
12       0        120                  0     0    0      0     0   
15       0      115.5                  0     0    0      0     0   
16       0          0   25.5 (explosion)     0    0      0     0   
17       0          0  12.65 (explosion)     0    0      0     0   
18       0          0                 11     0    0      0     0   
19     500          0                  0     0    0      0     0   
20     500          0                  0     0    0      0     0   
21      11          0                  0     0    0      0     0   
22      12          0                  0     0    0      0     0   
23  1.1(6)          0                  0     0    0      0     0   
24     2.5          0                  0     0    0      0     0   
25    28.5          0                  0     0    0      0     0   

            explosion  
0                   0  
1                   0  
2                   0  
3                   0  
5                   0  
6                   0  
10                  0  
11                  0  
12                  0  
15                  0  
16   25.5 (radiation)  
17  12.65 (radiation)  
18     11 (radiation)  
19                  0  
20                  0  
21                  0  
22                  0  
23                  0  
24                  0  
25                  0

Отлично, все верно. Теперь прибавим к каждой колонке префикс

In [57]:
split_bomb_damage_sec = split_bomb_damage_sec.add_suffix('_sec')

In [58]:
split_bomb_damage_sec['radiation_sec'] = split_bomb_damage_sec['radiation_sec'].apply(lambda x: x.replace(" (explosion)", ''))

In [59]:
split_bomb_damage_sec['explosion_sec'] = split_bomb_damage_sec['explosion_sec'].apply(lambda x: x.replace(" (radiation)", ''))

In [60]:
split_bomb_damage_sec.head()

damage_sec electrical_sec radiation_sec frost_sec fire_sec poison_sec  \
0        6.8              0             0         0        0          0   
1       71.5              0             0         0        0          0   
2       14.4              0             0         0        0          0   
3       82.8              0             0         0        0          0   
5        165              0             0         0        0          0   

  blood_sec explosion_sec  
0         0             0  
1         0             0  
2         0             0  
3         0             0  
5         0             0

В столбце "damage_sec" имеются такие значения как 1.1(6), отбросим значения в скобках

In [61]:
split_bomb_damage_sec['damage_sec'] = split_bomb_damage_sec['damage_sec'].apply(lambda x: x.split('('))

In [62]:
split_bomb_damage_sec['damage_sec'] = split_bomb_damage_sec['damage_sec'].apply(lambda x: x[0] if len(x) == 2 else x)

In [63]:
split_bomb_damage_sec = split_bomb_damage_sec.applymap(lambda x: x[0] if isinstance(x, list) else x)

In [64]:
split_bomb_damage_sec.head()

damage_sec electrical_sec radiation_sec frost_sec fire_sec poison_sec  \
0        6.8              0             0         0        0          0   
1       71.5              0             0         0        0          0   
2       14.4              0             0         0        0          0   
3       82.8              0             0         0        0          0   
5        165              0             0         0        0          0   

  blood_sec explosion_sec  
0         0             0  
1         0             0  
2         0             0  
3         0             0  
5         0             0

In [65]:
ranged_weapon = pd.concat([ranged_weapon,split_bomb_damage,split_bomb_damage_sec],axis = 1)

In [66]:
ranged_weapon = ranged_weapon.drop(["Урон","Урон_Секунду"], axis = 1)

In [67]:
list_float = ['Дальность','Точность','Магазин_Патронов','Вес','Цена']
list_floats = list(split_bomb_damage.columns.values) + list(split_bomb_damage_sec.columns.values) + list_float 

In [68]:
for list_float in list_floats:
    ranged_weapon[list_float] = ranged_weapon[list_float].apply(lambda x: str(x).replace(",", '.'))

In [69]:
ranged_weapon['Магазин_Патронов'] = ranged_weapon['Магазин_Патронов'].apply(lambda x: str(x).replace("∞", '100000'))

In [70]:
ranged_weapon.head(20)

Unnamed: 0                                Название Тип_Оружия  \
0            0                Гладкоствольный пистолет  Пистолеты   
1            1     Гладкоствольный с затвором пистолет  Пистолеты   
2            2  Гладкоствольный с патронником пистолет  Пистолеты   
3            3                          10-мм пистолет  Пистолеты   
5            5                               Спаситель  Пистолеты   
6            6                            .44 пистолет  Пистолеты   
10          10                    Ковбойский револьвер  Пистолеты   
11          11                   Институтский пистолет  Пистолеты   
12          12                       Лазерный пистолет  Пистолеты   
15          15                     Плазменный пистолет  Пистолеты   
16          16                             Гамма-пушка  Пистолеты   
17          17                          Дзета-пистолет  Пистолеты   
18          18                        Пистолет Лоренцо  Пистолеты   
19          19                      Внеземной пистолет  Пистолеты   
20          20                Бластер Чужих — Хабологи  Пистолеты   
21          21              Головной лазер штурмотрона  Пистолеты   
22          22                 Гладкоствольный карабин   Карабины   
23          23      Гладкоствольный c затвором карабин   Карабины   
24          24   Гладкоствольный с патронником карабин   Карабины   
25          25                   Двуствольный дробовик   Карабины   

                         Патроны Дальность Точность Магазин_Патронов   Вес  \
0                     Патрон .38      83.0     57.0               12   2.3   
1                    Патрон .308      95.0     70.0                6   4.0   
2                     Патрон .45      83.0     61.0                6   4.2   
3                   10-мм патрон      83.0     60.0               12   3.5   
5                   10-мм патрон      53.0     67.0               16   3.6   
6                     Патрон .44     119.0     66.0                6   5.2   
10                    Патрон .44     119.0     66.0                6   5.2   
11               Ядерная батарея      71.0     70.0               27   3.9   
12               Ядерная батарея      71.0     71.0               30   3.9   
15                  Заряд плазмы     149.0    142.0               30   3.9   
16        Патрон для гамма-пушки     149.0     69.0                8   3.0   
17        Патрон для гамма-пушки     149.0     69.0                8   3.1   
18        Патрон для гамма-пушки     254.0     69.0                8   4.5   
19  Боеприпас для бластера Чужих     149.0     73.0               42   2.5   
20  Боеприпас для бластера Чужих     119.0     73.0               42   2.5   
21               Ядерная батарея      93.0     60.0                1     8   
22                    Патрон .38      83.0     68.0               12   2.9   
23                   Патрон .308      95.0     70.0                6   4.0   
24                    Патрон .45      83.0     70.0                6   5.3   
25              Патрон дробовика      71.0     34.0                2  10.5   

      Цена damage  ... blood explosion damage_sec electrical_sec  \
0     20.0     13  ...     0         0        6.8              0   
1     36.0     35  ...     0         0       71.5              0   
2     25.0     25  ...     0         0       14.4              0   
3     50.0     18  ...     0         0       82.8              0   
5    149.0     27  ...     0         0        165              0   
6     99.0     48  ...     0         0       28.8              0   
10    99.0     60  ...     0         0          0              0   
11    50.0      0  ...     0         0          0             99   
12    69.0      0  ...     0         0          0            120   
15   123.0     25  ...     0         0          0          115.5   
16   156.0      0  ...     0        23          0              0   
17   156.0      0  ...     0      11.5          0              0   
18   753.0      0  ...     0   

In [71]:
ranged_weapon[list_floats] = ranged_weapon[list_floats].astype(float)

In [72]:
ranged_weapon.head(20)

Unnamed: 0                                Название Тип_Оружия  \
0            0                Гладкоствольный пистолет  Пистолеты   
1            1     Гладкоствольный с затвором пистолет  Пистолеты   
2            2  Гладкоствольный с патронником пистолет  Пистолеты   
3            3                          10-мм пистолет  Пистолеты   
5            5                               Спаситель  Пистолеты   
6            6                            .44 пистолет  Пистолеты   
10          10                    Ковбойский револьвер  Пистолеты   
11          11                   Институтский пистолет  Пистолеты   
12          12                       Лазерный пистолет  Пистолеты   
15          15                     Плазменный пистолет  Пистолеты   
16          16                             Гамма-пушка  Пистолеты   
17          17                          Дзета-пистолет  Пистолеты   
18          18                        Пистолет Лоренцо  Пистолеты   
19          19                      Внеземной пистолет  Пистолеты   
20          20                Бластер Чужих — Хабологи  Пистолеты   
21          21              Головной лазер штурмотрона  Пистолеты   
22          22                 Гладкоствольный карабин   Карабины   
23          23      Гладкоствольный c затвором карабин   Карабины   
24          24   Гладкоствольный с патронником карабин   Карабины   
25          25                   Двуствольный дробовик   Карабины   

                         Патроны  Дальность  Точность  Магазин_Патронов   Вес  \
0                     Патрон .38       83.0      57.0              12.0   2.3   
1                    Патрон .308       95.0      70.0               6.0   4.0   
2                     Патрон .45       83.0      61.0               6.0   4.2   
3                   10-мм патрон       83.0      60.0              12.0   3.5   
5                   10-мм патрон       53.0      67.0              16.0   3.6   
6                     Патрон .44      119.0      66.0               6.0   5.2   
10                    Патрон .44      119.0      66.0               6.0   5.2   
11               Ядерная батарея       71.0      70.0              27.0   3.9   
12               Ядерная батарея       71.0      71.0              30.0   3.9   
15                  Заряд плазмы      149.0     142.0              30.0   3.9   
16        Патрон для гамма-пушки      149.0      69.0               8.0   3.0   
17        Патрон для гамма-пушки      149.0      69.0               8.0   3.1   
18        Патрон для гамма-пушки      254.0      69.0               8.0   4.5   
19  Боеприпас для бластера Чужих      149.0      73.0              42.0   2.5   
20  Боеприпас для бластера Чужих      119.0      73.0              42.0   2.5   
21               Ядерная батарея       93.0      60.0               1.0   8.0   
22                    Патрон .38       83.0      68.0              12.0   2.9   
23                   Патрон .308       95.0      70.0               6.0   4.0   
24                    Патрон .45       83.0      70.0               6.0   5.3   
25              Патрон дробовика       71.0      34.0               2.0  10.5   

      Цена  damage  ...  blood  explosion  damage_sec  electrical_sec  \
0     20.0    13.0  ...    0.0        0.0         6.8             0.0   
1     36.0    35.0  ...    0.0        0.0        71.5             0.0   
2     25.0    25.0  ...    0.0        0.0        14.4             0.0   
3     50.0    18.0  ...    0.0        0.0        82.8             0.0   
5    149.0    27.0  ...    0.0        0.0       165.0             0.0   
6     99.0    48.0  ...    0.0        0.0        28.8             0.0   
10    99.0    60.0  ...    0.0        0.0         0.0             0.0   
11    50.0     0.0  ...    0.0        0.0         0.0            99.0   
12    69.0     0.0  ...    0.0        0.0         0.0           120.0   
15   123.0    25.0  ...    0.0        0.0         0.0           115.5   
16   156.0     0.0  ...    0.0       23.0        

In [73]:
ranged_weapon.dtypes

Unnamed: 0            int64
Название             object
Тип_Оружия           object
Патроны              object
Дальность           float64
Точность            float64
Магазин_Патронов    float64
Вес                 float64
Цена                float64
damage              float64
electrical          float64
radiation           float64
frost               float64
fire                float64
poison              float64
blood               float64
explosion           float64
damage_sec          float64
electrical_sec      float64
radiation_sec       float64
frost_sec           float64
fire_sec            float64
poison_sec          float64
blood_sec           float64
explosion_sec       float64
dtype: object

Привели таблицу к рабочему виду

## Таблица melee

In [74]:
melee

Unnamed: 0                              Название Тип_Оружия  \
0            0                          Удар боксёра   Кулачное   
1            1                                Кастет   Кулачное   
2            2                         Крюк для мяса   Кулачное   
3            3                        Мясницкий крюк   Кулачное   
4            4                        Силовой кастет   Кулачное   
5            5                 Перчатка когтя смерти   Кулачное   
6            6                          Выкидной нож   Холодное   
7            7                            Боевой нож   Холодное   
8            8                        Клинок Пикмана   Холодное   
9            9                  Клинок революционера   Холодное   
10          10                       Меч Шема Драуна   Холодное   
11          11                Меч китайского офицера   Холодное   
12          12                   Месть генерала Чжао   Холодное   
13          13                              Меч Цзао   Холодное   
14          14                             Меч чудес   Холодное   
15          15                                 Тесак   Холодное   
16          16                            Зуб Кремва   Холодное   
17          17                               Костыль   Холодное   
18          18                                 Валик   Холодное   
19          19                     Бейсбольный биток   Холодное   
20          20  Бейсбольный биток Мировой серии 2076   Холодное   
21          21                   Роквилльский боксёр   Холодное   
22          22                            Фенсбастер   Холодное   
23          23                 Блестящая дубина Цито   Холодное   
24          24                                   Кий   Холодное   
25          25                                  Дрын   Холодное   
26          26                                 Ломик   Холодное   
27          27                       Свинцовый стояк   Холодное   
28          28                           Потрошитель   Холодное   
29          29                          Страшный суд   Холодное   
30          30                                  Жнец   Холодное   
31          31                                  Прут   Холодное   
32          32                        Разводной ключ   Холодное   
33          33                          Большой Джим   Холодное   
34          34                                 Молот   Холодное   
35          35                             Шиш-кебаб   Холодное   
36          36                            Супермолот   Холодное   
37          37                        Топор Грогнака   Холодное   
38          38                    Атомное правосудие   Холодное   
39          39                    Клинок штурмотрона   Холодное   
40          40   Циркулярный резак Мистера Помощника   Холодное   
41          41                                 Багор   Холодное   
42          42           Кровопускатель (Far Harbor)   Холодное   
43          43                      Ловушка для рыбы   Холодное   
44          44                           Нож Адептов   Холодное   
45          45                              Горлорез   Холодное   
46          46                       Молот для комми   Холодное   

                        Урон Урон_Секунду  Очки_Действия  Вес  Цена  
0                          9         10.2           25.0    1    10  
1                         10          9.4           25.0  0.5    10  
2                         20          NaN            NaN    2    32  
3                         20          NaN            NaN    2    32  
4                         20         23.8           35.0    4   100  
5                         25         29.8           40.0   10    75  
6                          8       Быстр.           20.0    1    20  
7                          9       Быстр.           20.0    1    25  
8               9+25 (blood)       Быстр.           20.0    1   143  
9                         16       Средн.           35.0    3    50  
10          16+9 

In [75]:
melee.isna().mean()

Unnamed: 0       0.000000
Название         0.000000
Тип_Оружия       0.000000
Урон             0.000000
Урон_Секунду     0.042553
Очки_Действия    0.042553
Вес              0.000000
Цена             0.000000
dtype: float64

**Пропуски в столбце "Урон_Секунду"**

In [76]:
melee[melee['Урон_Секунду'].isna()]

Unnamed: 0        Название Тип_Оружия Урон Урон_Секунду  Очки_Действия Вес  \
2           2   Крюк для мяса   Кулачное   20          NaN            NaN   2   
3           3  Мясницкий крюк   Кулачное   20          NaN            NaN   2   

   Цена  
2    32  
3    32

Заполним урон в секунду 0, т.к. если урон не указан, значит за 1 секунду не наносится урон. Очки действия укажем 0, тк они не указаны

In [77]:
melee['Урон_Секунду'] = melee['Урон_Секунду'].fillna(0)

In [78]:
melee['Очки_Действия'] = melee['Очки_Действия'].fillna(0)

В столбце "Урон_Секунду" есть как значения float так и str, разделим эту таблицу на две

In [79]:
melee_float = melee[melee["Урон_Секунду"].apply(lambda x: pattern.search(str(x)) is None)]

In [80]:
melee_str = melee[melee["Урон_Секунду"].apply(lambda x: pattern.search(str(x)) is not None)]

Приведем таблицу melee_float к нужному виду

In [81]:
melee_float[list(melee_float.columns)[3:]] = melee_float[list(melee_float.columns)[3:]].astype(float)

In [82]:
melee_float = melee_float.rename(columns = {'Урон' : 'damage'})

In [83]:
melee_float.dtypes

Unnamed: 0         int64
Название          object
Тип_Оружия        object
damage           float64
Урон_Секунду     float64
Очки_Действия    float64
Вес              float64
Цена             float64
dtype: object

Таблица melee_float готова. Теперь отсортируем таблицу melee_str

In [84]:
melee_str.head()

Unnamed: 0              Название Тип_Оружия              Урон  \
6            6          Выкидной нож   Холодное                 8   
7            7            Боевой нож   Холодное                 9   
8            8        Клинок Пикмана   Холодное      9+25 (blood)   
9            9  Клинок революционера   Холодное                16   
10          10       Меч Шема Драуна   Холодное  16+9 (radiation)   

   Урон_Секунду  Очки_Действия Вес  Цена  
6        Быстр.           20.0   1    20  
7        Быстр.           20.0   1    25  
8        Быстр.           20.0   1   143  
9        Средн.           35.0   3    50  
10       Средн.           35.0   3   250

In [85]:
melee_str['Вес'] = melee_str['Вес'].apply(lambda x: x.replace(",", '.'))

In [86]:
split_melee_str_damage = damage_sorting(melee_str["Урон"])

In [87]:
split_melee_str_damage['electrical'] = split_melee_str_damage['electrical'].apply(lambda x: x.replace("(electrical)", ''))

In [88]:
split_melee_str_damage['blood'] = split_melee_str_damage['blood'].apply(lambda x: x.replace(" (poison)", ''))

In [89]:
split_melee_str_damage['blood'] = split_melee_str_damage['blood'].apply(lambda x: x.replace("(blood)", ''))

In [90]:
melee_str = pd.concat([melee_str,split_melee_str_damage],axis = 1)

In [91]:
melee_str = melee_str.drop('Урон', axis = 1)

In [92]:
colmn = list(split_melee_str_damage.columns) + ['Очки_Действия','Вес','Цена']

In [93]:
melee_str[colmn] = melee_str[colmn].astype(float)

In [94]:
melee_str.dtypes

Unnamed: 0         int64
Название          object
Тип_Оружия        object
Урон_Секунду      object
Очки_Действия    float64
Вес              float64
Цена             float64
damage           float64
electrical       float64
radiation        float64
frost            float64
fire             float64
poison           float64
blood            float64
explosion        float64
dtype: object

Закончили с таблицей melee_str. Всего у нас 4 таблицы:
+ bomb - характеристики взрывчатки 
+ ranged_weapon - характеристики дальнобойного оружия 
+ melee_float - таблица ближнего боя (с числовым типом УВС)
+ melee_str - таблица ближнего боя (с строковым типом УВС)

## Таблица creatures

In [95]:
creatures.head(20)

Unnamed: 0        Вид                                    Название Уровень  \
0            0  Гуманоиды                                  Дикий гуль       3   
1            1  Гуманоиды                          Дикий гуль-бродяга       9   
2            2  Гуманоиды                                Розовый гуль      14   
3            3  Гуманоиды                          Дикий гуль-охотник      15   
4            4  Гуманоиды                       Дикий гуль-пожиратель      22   
5            5  Гуманоиды                          Усохший дикий гуль      32   
6            6  Гуманоиды                     Гангренозный дикий гуль      42   
7            7  Гуманоиды                          Гниющий дикий гуль      52   
8            8  Гуманоиды                       Обугленный дикий гуль     62+   
9            9  Гуманоиды                                  Светящийся      22   
10          10  Гуманоиды                           Гнилой светящийся      40   
11          11  Гуманоиды                      Раздувшийся светящийся     58+   
12          12  Гуманоиды                 Туманный гуль (Far Harbor)        3   
13          13  Гуманоиды             Раскрашенный гуль (Nuka-World)       22   
14          14  Гуманоиды  Обожжённый раскрашенный гуль (Nuka-World)       52   
15          15  Гуманоиды                                 Супермутант      10   
16          16  Гуманоиды                       Супермутант-камикадзе   10—42   
17          17  Гуманоиды                       Супермутант-разведчик       ?   
18          18  Гуманоиды                        Супермутант-налётчик       ?   
19          19  Гуманоиды                          Супермутант-бестия      22   

   Здоровье Восприятие Очки_Опыта Сопротивление_ФизУрон  \
0        35          2          7                    10   
1        70          2         13                    15   
2        10          2         22                    15   
3        80          2         22                    30   
4       140          2         22                    40   
5       170          3         47                    50   
6       200          3         61                    60   
7       385          4         76                    70   
8      505+          5         90                    85   
9       400          3         32                    40   
10      650          4         57                    75   
11     785+          5         83                    90   
12       35          2          7                    10   
13      140          2         32                    40   
14      385          4         76                    70   
15      120          7          7                    10   
16  120—600       7—10       7—54                 17—72   
17        ?          ?          ?                     ?   
18        ?          ?          ?                     ?   
19        ?          ?          ?                     ?   

   Сопротивление_ЭнергоУрон Сопротивление_РадУрон Сопротивление_РадУронду  \
0                        20                  1000                      50   
1                        30                  1000                      50   
2                        15                  1000                     200   
3                        45                  1000                      50   
4                        60                  1000                      50   
5                        75                  1000                      50   
6                        90                  1000                      50   
7                       110                  1000                     225   
8                       140                  1000                     250   
9                        80                  1000                     200   
10                      150                  1000                     200   
11                      175                  1000                     225   
12                       20                  1000            

In [96]:
creatures = creatures.rename(columns = {'Сопротивление_РадУронду':'Сопротивление_ЯдУрону'})

Удалим те строки, где много присутствует знак "?" и названия 'Name?'

In [97]:
colmn = creatures.columns[3:11]

In [98]:
query = ((creatures[colmn[1]] != '?') | (creatures[colmn[2]] != '?') |
        (creatures[colmn[3]] != '?') | (creatures[colmn[4]] != '?') | (creatures[colmn[5]] != '?'))

In [99]:
creatures = creatures[query]

In [100]:
creatures = creatures[creatures['Название'] != 'Name?']

Избавимся от "+" в ячейках

In [101]:
columns = list(creatures.dtypes[creatures.dtypes == 'object'].index)
for column in columns:
    creatures[column] = creatures[column].apply(lambda x: x.replace('+',''))

Сделаем средние значения для "Супермутант-камикадзе" с разбросаными характеристиками

In [102]:
creatures[creatures['Уровень'].apply(lambda x: pattern.search(x) is not None)]

Empty DataFrame
Columns: [Unnamed: 0, Вид, Название, Уровень, Здоровье, Восприятие, Очки_Опыта, Сопротивление_ФизУрон, Сопротивление_ЭнергоУрон, Сопротивление_РадУрон, Сопротивление_ЯдУрону, Агрессия, Уверенность, Помощь]
Index: []

In [103]:
regex = "—" 
pattern = re.compile(regex)

blanks = list(creatures[creatures['Уровень'].apply(lambda x: pattern.search(x) is not None)].iloc[0])
blank = [blank.split("—") for blank in blanks if '—' in str(blank)]
stat = [str([np.mean([list(map(int, blank[i])) for i in range(len(blank))][i]) for i in range(len(blank))][i]) for i in range(len(blank))]
stat.insert(-1, '100000')
super_mutant = list(creatures[creatures['Уровень'].apply(lambda x: pattern.search(x) is not None)].iloc[0])[:3] + stat + [0,0,0]

In [104]:
creatures = creatures.drop(labels = [16],axis = 0)

In [105]:
creatures.loc[ len(creatures.index )] = super_mutant

In [106]:
creatures[creatures['Название'] == 'Супермутант-камикадзе']

Unnamed: 0        Вид               Название Уровень Здоровье Восприятие  \
202          16  Гуманоиды  Супермутант-камикадзе    26.0    360.0        8.5   

    Очки_Опыта Сопротивление_ФизУрон Сопротивление_ЭнергоУрон  \
202       30.5                  44.5                     27.5   

    Сопротивление_РадУрон Сопротивление_ЯдУрону  Агрессия  Уверенность  Помощь  
202                100000                  17.5         0            0       0

Заменим знак вопроса на 0 и уберем все остальные ненужные знаки

In [107]:
regex_ru = "[а-яА-Я]+" 
pattern_ru = re.compile(regex_ru)

In [108]:
columns = creatures.columns[3:11]
for column in columns:
    creatures[column] = creatures[column].str.split('/').apply(lambda x: x[0] if (len(x) == 2) else x)
    creatures = creatures.applymap(lambda x: x[0] if isinstance(x, list) else x)
    creatures[column] = creatures[column].str.split('(').apply(lambda x: x[0] if (len(x) == 2) else x)
    creatures = creatures.applymap(lambda x: x[0] if isinstance(x, list) else x)
    creatures[column] = creatures[column].apply(lambda x: x.replace('?','0'))
    creatures[column] = creatures[column].apply(lambda x: x.replace('100\xa0%','100000'))
    creatures[column] = creatures[column].apply(lambda x: x.replace('100%','100000'))
    creatures[column] = creatures[column].apply(lambda x: '100000' if (pattern_ru.search(x) is not None) else x )

In [109]:
 creatures.dtypes

Unnamed: 0                   int64
Вид                         object
Название                    object
Уровень                     object
Здоровье                    object
Восприятие                  object
Очки_Опыта                  object
Сопротивление_ФизУрон       object
Сопротивление_ЭнергоУрон    object
Сопротивление_РадУрон       object
Сопротивление_ЯдУрону       object
Агрессия                     int64
Уверенность                  int64
Помощь                       int64
dtype: object

Приведем столбцы к нужному типу данных

In [110]:
creatures.iloc[:,3:] = creatures.iloc[:,3:].astype(float)

In [111]:
creatures.dtypes

Unnamed: 0                    int64
Вид                          object
Название                     object
Уровень                     float64
Здоровье                    float64
Восприятие                  float64
Очки_Опыта                  float64
Сопротивление_ФизУрон       float64
Сопротивление_ЭнергоУрон    float64
Сопротивление_РадУрон       float64
Сопротивление_ЯдУрону       float64
Агрессия                    float64
Уверенность                 float64
Помощь                      float64
dtype: object

In [112]:
creatures.head(20)

Unnamed: 0        Вид                                    Название  \
0            0  Гуманоиды                                  Дикий гуль   
1            1  Гуманоиды                          Дикий гуль-бродяга   
2            2  Гуманоиды                                Розовый гуль   
3            3  Гуманоиды                          Дикий гуль-охотник   
4            4  Гуманоиды                       Дикий гуль-пожиратель   
5            5  Гуманоиды                          Усохший дикий гуль   
6            6  Гуманоиды                     Гангренозный дикий гуль   
7            7  Гуманоиды                          Гниющий дикий гуль   
8            8  Гуманоиды                       Обугленный дикий гуль   
9            9  Гуманоиды                                  Светящийся   
10          10  Гуманоиды                           Гнилой светящийся   
11          11  Гуманоиды                      Раздувшийся светящийся   
12          12  Гуманоиды                 Туманный гуль (Far Harbor)    
13          13  Гуманоиды             Раскрашенный гуль (Nuka-World)    
14          14  Гуманоиды  Обожжённый раскрашенный гуль (Nuka-World)    
15          15  Гуманоиды                                 Супермутант   
25          25  Гуманоиды                           Супермутант-вождь   
27          27  Гуманоиды                                       Чужой   
28          28  Гуманоиды                       Инопланетянин(fo4cc)    
29          29  Гуманоиды                       Инопланетянин(fo4cc)    

    Уровень  Здоровье  Восприятие  Очки_Опыта  Сопротивление_ФизУрон  \
0       3.0      35.0         2.0         7.0                   10.0   
1       9.0      70.0         2.0        13.0                   15.0   
2      14.0      10.0         2.0        22.0                   15.0   
3      15.0      80.0         2.0        22.0                   30.0   
4      22.0     140.0         2.0        22.0                   40.0   
5      32.0     170.0         3.0        47.0                   50.0   
6      42.0     200.0         3.0        61.0                   60.0   
7      52.0     385.0         4.0        76.0                   70.0   
8      62.0     505.0         5.0        90.0                   85.0   
9      22.0     400.0         3.0        32.0                   40.0   
10     40.0     650.0         4.0        57.0                   75.0   
11     58.0     785.0         5.0        83.0                   90.0   
12      3.0      35.0         2.0         7.0                   10.0   
13     22.0     140.0         2.0        32.0                   40.0   
14     52.0     385.0         4.0        76.0                   70.0   
15     10.0     120.0         7.0         7.0                   10.0   
25     68.0    1535.0         8.0         0.0                  135.0   
27      1.0      50.0         8.0        25.0                    0.0   
28      1.0      50.0         8.0        25.0                    0.0   
29      1.0      50.0         8.0        25.0                    0.0   

    Сопротивление_ЭнергоУрон  Сопротивление_РадУрон  Сопротивление_ЯдУрону  \
0                       20.0                 1000.0                   50.0   
1                       30.0                 1000.0                   50.0   
2                       15.0                 1000.0                  200.0   
3                       45.0                 1000.0                   50.0   
4                       60.0                 1000.0                   50.0   
5                       75.0                 1000.0                   50.0   
6                       90.0                 1000.0                   50.0   
7                      110.0                 1000.0                  225.0   
8                      140.0                 1000.0                  250.0   
9                       80.0                 1000.0                  200.0   
10                     150.0                 1000.0                  200.0   
11                   

Удалим существ с здоровьем равным 0

In [113]:
creatures = creatures[creatures['Здоровье'] != 0]

## Соединяем таблицы

In [114]:
all_weapon = pd.concat([ranged_weapon,bomb,melee_float, melee_str], axis = 0)

In [115]:
all_weapon = all_weapon.drop('Unnamed: 0', axis = 1).reset_index(drop=True)

In [116]:
all_weapon['index'] = all_weapon.index

In [117]:
creatures = creatures.drop('Unnamed: 0', axis = 1).reset_index(drop=True)

In [118]:
creatures['index'] = creatures.index

In [119]:
creatures = creatures.rename(columns = {"Название":"Название_существа"})

In [120]:
df = creatures.merge(all_weapon, how = 'cross')

In [121]:
df = df.fillna(0)

Присвоим для оружия бляжнего боя точность 100000

In [122]:
df.loc[df["Тип_Оружия"]=="Холодное"] = df.loc[df["Тип_Оружия"]=="Холодное"].replace({'Точность' : {0 : 10000}})

In [123]:
df.loc[df["Тип_Оружия"]=="Кулачное"] = df.loc[df["Тип_Оружия"]=="Кулачное"].replace({'Точность' : {0 : 10000}})

In [124]:
df.loc[df["Тип_Оружия"]=="Взрывчатка"] = df.loc[df["Тип_Оружия"]=="Взрывчатка"].replace({'Точность' : {0 : 10000}})

In [125]:
df.head(25)

Вид Название_существа  Уровень  Здоровье  Восприятие  Очки_Опыта  \
0   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
1   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
2   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
3   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
4   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
5   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
6   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
7   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
8   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
9   Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
10  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
11  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
12  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
13  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
14  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
15  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
16  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
17  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
18  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
19  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
20  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
21  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
22  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
23  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   
24  Гуманоиды        Дикий гуль      3.0      35.0         2.0         7.0   

    Сопротивление_ФизУрон  Сопротивление_ЭнергоУрон  Сопротивление_РадУрон  \
0                    10.0                      20.0                 1000.0   
1                    10.0                      20.0                 1000.0   
2                    10.0                      20.0                 1000.0   
3                    10.0                      20.0                 1000.0   
4                    10.0                      20.0                 1000.0   
5                    10.0                      20.0                 1000.0   
6                    10.0                      20.0                 1000.0   
7                    10.0                      20.0                 1000.0   
8                    10.0                      20.0                 1000.0   
9                    10.0                      20.0                 1000.0   
10                   10.0                      20.0                 1000.0   
11                   10.0                      20.0                 1000.0   
12                   10.0                      20.0                 1000.0   
13                   10.0                      20.0                 1000.0   
14                   10.0                      20.0                 1000.0   
15                   10.0                      20.0                 1000.0   
16                   10.0                      20.0                 1000.0   
17                   10.0                      20.0                 1000.0   
18                   10.0                      20.0                 1000.0   
19                   10.0                      20.0                 1000.0   
20                   10.0                      20.0                 1000.0   
21                   10.0                      20.0                 1000.0   
22                   10.0                      20.0                 1000.0   
23                   10.0                      20.0                 1000.0   
24                   10.0      

In [126]:
df.to_csv('creatures_weapons.csv', index = False)

# Построение модели предсказания

## Подсчет кол-во патронов для полного поражения цели

Построим модель предсказания **кол-во патронов** для убийства существа. Посчитаем это значение с помощью формул, а далее будем предсказывать это значение для тестовой выборки.  
Чтобы посчитать этот параметр отбросим большую часть столбцов в таблице, будем использовать только те, которые *непосредственно* и *наверняка* участвуют в подсчете параметра (кол-во патронов)  
В расчете формул будем непосредственно отталкиваться от информации наносимого урона в игре fallout 4

**Damage Resistance**  
Посмотрим как существа получают урон на основе характеристик сопротивления к урону. Для этого обратимся к статье на сайте [Fandom](https://fallout.fandom.com/wiki/Damage_Resistance#Fallout_4) в которой отлично расписана формула коэффициента нанесения урона от оружия. 

$$ DamageCoeff =  0.99 * \left(\frac{fPhysicalDamageFactor * Damage}{DamageResist} \right)^{0.365} $$
+ fPhysicalDamageFactor - коэффициент урона, в игре по умолчанию 0.15  
+ Damage - наносимый урон оружием
+ DamageResist - параметр сопротивлению к урону
+ DamageCoeff - процент наносимого урона от атаки оружием  
Данный коэффициент умножается на урон от оружия, это и будет наносимый урон здоровью существа

Для комбинированного типа повреждения (когда оружие может наносить как физический, так и радиционный урон и т.д.) одно попадание пули расценивается, как комбинированный урон от разных типов повреждения.

Для такого урона как radiation, fire, frost, blood не будем высчитывать дополнительные временные эффекты урона.  
Для вида урона fire, frost, blood, explosion нет параметров сопративления у существ, поэтому этот тип урона будем высчитывать как обычный физический

**Range affect damage**

Параметр "Дальность" говорит нам о расстоянии, после которого урон патрона снижается. Согласно статье на [Fandom](https://fallout.fandom.com/wiki/Range#Fallout_4):
> максимальное расстояние, на котором конкретное оружие будет наносить максимальный урон за пределами VATS и дополнительно влияет на шанс поразить цель при использовании VATS

Согласно странице сопротивления урону в [Fandom](https://fallout.fandom.com/wiki/Damage_Resistance#Fallout_4) , урон от оружия за пределами этого максимального диапазона уменьшит урон до 50% при удвоении диапазона
>RangeMultiсоставляет 1,0 при 100% дальности оружия или ближе, 0,5 при 200% WeaponRangeили более, и масштабируется от 1,0 до 0,5 по мере увеличения дальности от 100% до 200% от WeaponRange.

Таким образом, для облегчения расчетов будем считать, что расстояния выстрела всегда оптимальное и не переходит черту, тем самым при каждом выстреле будет наносится 100% урон

**Accuracy Fallout 4**

Точность в Fallout 4 влияет на разброс пуль согласно ответу одному пользователю на форуме [gamespot.com](https://gamefaqs.gamespot.com/boards/164594-fallout-4/72871814)
>Точность в большинстве игр типа fps относятся к разбросу выстрела после его выстрела, а также к разбросу выстрела от бедра большинство оружия с низкой точностью будет иметь огромное поле разброса, что означает, что даже если ваша прицельная сетка находится в центре врага, ваша пуля все еще может оказаться далеко слева от прицельной сетки. именно так точность используется в этой игре.

Из-за того, что стрельба ведется с оптимального расстояния, то процент промахов будет достаточно низким, поэтому не будем брать во внимания данную метрику

остальные метрики, мы не будем высчитывать т.к. 
1. Нет информации и формул, как эти показатели связаны друг с другом, например характеристика "Восприятия" у существ
2. Другие переменные трудно интепретировать, например "Агрессия", "Помощь" и т.д. данные характеристики влияют на контекст встречи с существом, поэтому отбросим их
3. Ненужные данные, например тип патронов и т.д.

In [127]:
columns = ['Название_существа', 'Здоровье','Сопротивление_ФизУрон', 'Сопротивление_ЭнергоУрон', 'Сопротивление_РадУрон',
           'Сопротивление_ЯдУрону', 'Название', 'Тип_Оружия', 'damage', 'electrical', 'radiation', 'frost', 'fire', 
           'poison', 'blood', 'explosion']

In [128]:
df = df[columns]

In [129]:
"""
Функция считает кол-во патронов необходимое для убиства существа
"""
def ammo_count(df,x):
    map_damage = {'damage' : 'Сопротивление_ФизУрон',
                 'fire' : 'Сопротивление_ФизУрон',
                 'frost' : 'Сопротивление_ФизУрон',
                 'blood' : 'Сопротивление_ФизУрон',
                 'explosion' : 'Сопротивление_ФизУрон',
                 'electrical' : 'Сопротивление_ЭнергоУрон',
                 'radiation' : 'Сопротивление_РадУрон',
                 'poison' : 'Сопротивление_ЯдУрону'}
    
    name_columns = list(map_damage.keys())
    
    dmg = []
    for column in name_columns:
        fPhysicalDamage = 0.15
        damage = df[column].loc[x] #номер строки
        DamageResist = df[map_damage[column]].loc[x]
        dmg.append(df[column].loc[x] * (((fPhysicalDamage * damage) / DamageResist) ** 0.365))
    dmg_one_ammo = np.sum(dmg)
    hp = df['Здоровье'].loc[x]
    ammo = 0
    while hp > 0:
        hp -= dmg_one_ammo
        ammo += 1
    return ammo

In [130]:
from tqdm import tqdm 
from IPython.display import clear_output

amount_ammo = []
[amount_ammo.append(ammo_count(df,i)) for i in tqdm(range(len(df)))]
clear_output()

In [131]:
df['Ammo_shots'] =  amount_ammo

In [132]:
df.head()

Название_существа  Здоровье  Сопротивление_ФизУрон  \
0        Дикий гуль      35.0                   10.0   
1        Дикий гуль      35.0                   10.0   
2        Дикий гуль      35.0                   10.0   
3        Дикий гуль      35.0                   10.0   
4        Дикий гуль      35.0                   10.0   

   Сопротивление_ЭнергоУрон  Сопротивление_РадУрон  Сопротивление_ЯдУрону  \
0                      20.0                 1000.0                   50.0   
1                      20.0                 1000.0                   50.0   
2                      20.0                 1000.0                   50.0   
3                      20.0                 1000.0                   50.0   
4                      20.0                 1000.0                   50.0   

                                 Название Тип_Оружия  damage  electrical  \
0                Гладкоствольный пистолет  Пистолеты    13.0         0.0   
1     Гладкоствольный с затвором пистолет  Пистолеты    35.0         0.0   
2  Гладкоствольный с патронником пистолет  Пистолеты    25.0         0.0   
3                          10-мм пистолет  Пистолеты    18.0         0.0   
4                               Спаситель  Пистолеты    27.0         0.0   

   radiation  frost  fire  poison  blood  explosion  Ammo_shots  
0        0.0    0.0   0.0     0.0    0.0        0.0           5  
1        0.0    0.0   0.0     0.0    0.0        0.0           2  
2        0.0    0.0   0.0     0.0    0.0        0.0           3  
3        0.0    0.0   0.0     0.0    0.0        0.0           4  
4        0.0    0.0   0.0     0.0    0.0        0.0           2

## Построем модель предсказания

### Сделаем Baseline

In [133]:
from sklearn.model_selection import train_test_split

In [134]:
! pip install catboost

In [135]:
train, test = train_test_split(df,train_size = 0.6, random_state=42)
val, test = train_test_split(test,train_size = 0.5, random_state=42)

In [136]:
from catboost import CatBoostRegressor

In [137]:
categorical_features_indices  = np.where(df.dtypes == np.object_)[0]
X = list(df)[:-1]
y = ['Ammo_shots']

In [138]:
from catboost import Pool

train_data = Pool(data=train[X],
                  label=train[y],
                  cat_features=categorical_features_indices
                 )

valid_data = Pool(data=val[X],
                  label=val[y],
                  cat_features=categorical_features_indices
                 )

In [139]:
params = {'verbose' : 100}

In [140]:
model = CatBoostRegressor(**params)

In [141]:
model.fit(train_data,eval_set=valid_data)

Learning rate set to 0.078225
0:	learn: 614.3471319	test: 847.3022812	best: 847.3022812 (0)	total: 163ms	remaining: 2m 43s
100:	learn: 121.7961935	test: 326.1234993	best: 326.1234993 (100)	total: 2.37s	remaining: 21.1s
200:	learn: 97.0783163	test: 301.5039125	best: 301.5039125 (200)	total: 4.38s	remaining: 17.4s
300:	learn: 74.9637560	test: 277.3075381	best: 277.3075381 (300)	total: 6.58s	remaining: 15.3s
400:	learn: 61.4515497	test: 267.1587232	best: 267.1587232 (400)	total: 8.83s	remaining: 13.2s
500:	learn: 52.8742147	test: 257.5154542	best: 257.5154542 (500)	total: 11.1s	remaining: 11s
600:	learn: 48.1402096	test: 256.0109090	best: 255.9973097 (596)	total: 13.5s	remaining: 8.94s
700:	learn: 41.8909122	test: 254.1532584	best: 253.9740349 (697)	total: 16.1s	remaining: 6.86s
800:	learn: 37.7229602	test: 249.8669802	best: 249.8216434 (789)	total: 18.4s	remaining: 4.56s
900:	learn: 34.6312213	test: 249.2160738	best: 249.2108815 (898)	total: 20.6s	remaining: 2.26s
999:	learn: 31.9134703	

In [142]:
from sklearn.metrics import mean_absolute_error,mean_absolute_percentage_error

In [143]:
def error(y_true,y_pred):
  print(mean_absolute_error(y_true,y_pred))
  print(mean_absolute_percentage_error(y_true,y_pred))

In [144]:
test['Ammo_shots_pred'] = model.predict(test[X])

In [145]:
error(test['Ammo_shots'], test['Ammo_shots_pred'])

15.37835953081639
3.040336686403053


Получили абсолютную ошибку в 15.3 - это значит, что настолько в среднем отклоняется наш предсказанный параметр от настоящего  
Процентная абсолютная ошибка - 3.04, что означает, на какой процент отклоняется в среднем наша предсказанная переменная от настоящей

### Подберем лучшие параметры через Hyperopt

In [146]:
!pip install hyperopt

In [147]:
from hyperopt import hp

In [148]:
ctb_reg_params = {
    'learning_rate':     hp.loguniform('learning_rate', -5, 0),
    'l2_leaf_reg':       hp.choice('l2_leaf_reg', np.arange(1, 8, 1)),
    'max_depth': hp.quniform('max_depth', 2, 10, 1),
    'n_estimators':  hp.quniform('n_estimators',  1500, 5000, 500),
    'random_seed' : 42,
    'eval_metric': hp.choice('eval_metric', ['RMSE','MAE','MAPE']),
    
    'cat_features': categorical_features_indices
}
ctb_fit_params = {
    'early_stopping_rounds': 100,
    'verbose': False
}
ctb_para = dict()
ctb_para['reg_params'] = ctb_reg_params
ctb_para['fit_params'] = ctb_fit_params
ctb_para['loss_func' ] = lambda y, pred: mean_absolute_percentage_error(y, pred) #здесь мы указываем функцию потерь

In [149]:
from hyperopt import fmin, tpe, STATUS_OK, STATUS_FAIL, Trials

In [150]:
x_train, x_val, y_train, y_val = train_test_split(df[X],df[y],train_size=0.8,random_state=42)

In [151]:
class HPOpt(object):

    def __init__(self, x_train, x_val, y_train, y_val):
        self.x_train = x_train
        self.x_val  = x_val
        self.y_train = y_train
        self.y_val  = y_val

    def process(self, fn_name, space, trials, algo, max_evals):
        fn = getattr(self, fn_name)
        try:
            result = fmin(fn=fn, space=space, algo=algo, max_evals=max_evals, trials=trials)
        except Exception as e:
            return {'status': STATUS_FAIL,
                    'exception': str(e)}
        return result, trials
    
    def ctb_reg(self, para):
        reg = CatBoostRegressor(**para['reg_params'])
        return self.train_reg(reg, para)

    def train_reg(self, reg, para):
        reg.fit(self.x_train, self.y_train,
                eval_set=[(self.x_train, self.y_train), (self.x_val, self.y_val)],**para['fit_params'])
        pred = reg.predict(self.x_val)
        loss = para['loss_func'](self.y_val, pred)
        return {'loss': loss, 'status': STATUS_OK}

In [152]:
obj = HPOpt(x_train, x_val, y_train, y_val)

ctb_opt = obj.process(fn_name='ctb_reg', space=ctb_para, trials=Trials(), algo=tpe.suggest, max_evals=100)

100%|██████████| 100/100 [3:57:04<00:00, 142.25s/trial, best loss: 0.3478310548730868]  


In [153]:
ctb_opt

({'eval_metric': 1,
  'l2_leaf_reg': 0,
  'learning_rate': 0.013333789619679086,
  'max_depth': 10.0,
  'n_estimators': 5000.0},
 <hyperopt.base.Trials at 0x2e5046e1370>)

### Построим модель с найденными параметрами

In [171]:
train, test = train_test_split(df,train_size = 0.6, random_state=42)
val, test = train_test_split(test,train_size = 0.5, random_state=42)

In [172]:
train_data = Pool(data=train[X],
                  label=train[y],
                  cat_features=categorical_features_indices
                 )

valid_data = Pool(data=val[X],
                  label=val[y],
                  cat_features=categorical_features_indices
                 )

In [173]:
params = {'eval_metric': 'MAE',
          'l2_leaf_reg': 0,
          'learning_rate': 0.013333789619679086,
          'max_depth': 10.0,
          'n_estimators': 5000.0}

In [174]:
model = CatBoostRegressor(verbose=200,**params)

In [175]:
model.fit(train_data,eval_set=valid_data)

0:	learn: 128.1209612	test: 144.9066587	best: 144.9066587 (0)	total: 71.7ms	remaining: 5m 58s
200:	learn: 20.7261184	test: 30.9643909	best: 30.9643909 (200)	total: 10.5s	remaining: 4m 10s
400:	learn: 7.3906966	test: 16.1032430	best: 16.1032430 (400)	total: 20.7s	remaining: 3m 57s
600:	learn: 5.3315400	test: 13.9666858	best: 13.9666858 (600)	total: 30.7s	remaining: 3m 44s
800:	learn: 4.2538559	test: 12.9918718	best: 12.9918718 (800)	total: 41.1s	remaining: 3m 35s
1000:	learn: 3.5002060	test: 12.3920780	best: 12.3920780 (1000)	total: 50.9s	remaining: 3m 23s
1200:	learn: 2.9860521	test: 12.0062321	best: 12.0062321 (1200)	total: 1m	remaining: 3m 12s
1400:	learn: 2.5771293	test: 11.7523033	best: 11.7523033 (1400)	total: 1m 11s	remaining: 3m 2s
1600:	learn: 2.2539480	test: 11.5441128	best: 11.5441128 (1600)	total: 1m 20s	remaining: 2m 51s
1800:	learn: 2.0101245	test: 11.3873108	best: 11.3873108 (1800)	total: 1m 31s	remaining: 2m 42s
2000:	learn: 1.8151162	test: 11.2791444	best: 11.2791444 (2

In [176]:
test['Ammo_shots_pred_hyperparameters'] = model.predict(test[X])

In [177]:
error(test['Ammo_shots'], test['Ammo_shots_pred_hyperparameters'])

9.809487758462977
0.35590523960648923


С подбором гиперпараметров наши результаты предсказания улучшились

### Сделаем кросс-валидацию на всех данных

In [178]:
train_full = pd.concat([train, val])

In [179]:
train_full_data = Pool(data=train_full[X],
                  label=train_full[y],
                  cat_features=categorical_features_indices
                 )

In [180]:
from catboost import cv

In [181]:
params = {'eval_metric': 'MAE',
          'l2_leaf_reg': 0,
          'learning_rate': 0.013333789619679086,
          'max_depth': 10.0,
          'n_estimators': 5000.0,
          "loss_function" : 'MAE' }

In [182]:
cv_date = cv(
    params = params,
    pool = train_full_data,
    fold_count = 5,
    shuffle=True,
    partition_random_seed=0,
    verbose=False,
    plot=True
)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

Training on fold [0/5]

bestTest = 22.62972973
bestIteration = 4998

Training on fold [1/5]

bestTest = 9.900033337
bestIteration = 4992

Training on fold [2/5]

bestTest = 13.83927477
bestIteration = 4999

Training on fold [3/5]

bestTest = 14.65254188
bestIteration = 4596

Training on fold [4/5]

bestTest = 12.56934887
bestIteration = 4650



In [183]:
cv_date

iterations  test-MAE-mean  test-MAE-std  train-MAE-mean  train-MAE-std
0              0      89.621675     13.058337       89.622465       3.266610
1              1      89.433490     13.047955       89.433782       3.275797
2              2      89.284466     13.001071       89.285727       3.318702
3              3      89.160496     13.029521       89.158443       3.299586
4              4      88.958697     12.933854       88.962611       3.385620
...          ...            ...           ...             ...            ...
4995        4995      14.753908      4.770328        8.499149       0.947977
4996        4996      14.753772      4.770474        8.498325       0.946637
4997        4997      14.753648      4.770088        8.495977       0.946971
4998        4998      14.753390      4.770149        8.495588       0.946727
4999        4999      14.753390      4.770157        8.495194       0.946715

[5000 rows x 5 columns]

In [184]:
cv_date[cv_date['test-MAE-mean'] == cv_date['test-MAE-mean'].min()].values[0][0]

4957.0

In [185]:
n_iters = cv_date[cv_date['test-MAE-mean'] == cv_date['test-MAE-mean'].min()].values[0][0]

### Сделаем финальное предсказание

In [186]:
params = {'eval_metric': 'MAE',
          'l2_leaf_reg': 0,
          'learning_rate': 0.013333789619679086,
          'max_depth': 10.0,
          'n_estimators': n_iters}

In [187]:
model = CatBoostRegressor(verbose=500,**params)

In [188]:
model.fit(train_data,eval_set=valid_data)

0:	learn: 128.1209612	test: 144.9066587	best: 144.9066587 (0)	total: 69.8ms	remaining: 5m 45s
500:	learn: 6.1190998	test: 14.7183282	best: 14.7183282 (500)	total: 24.5s	remaining: 3m 37s
1000:	learn: 3.5002060	test: 12.3920780	best: 12.3920780 (1000)	total: 49.8s	remaining: 3m 16s
1500:	learn: 2.4050788	test: 11.6375672	best: 11.6375672 (1500)	total: 1m 14s	remaining: 2m 50s
2000:	learn: 1.8151162	test: 11.2791444	best: 11.2791444 (2000)	total: 1m 38s	remaining: 2m 25s
2500:	learn: 1.4353449	test: 11.0713800	best: 11.0713800 (2500)	total: 2m 4s	remaining: 2m 2s
3000:	learn: 1.1827668	test: 10.9493833	best: 10.9488171 (2996)	total: 2m 28s	remaining: 1m 36s
3500:	learn: 0.9907524	test: 10.8669237	best: 10.8668202 (3499)	total: 2m 53s	remaining: 1m 12s
4000:	learn: 0.8475847	test: 10.8050810	best: 10.8050810 (4000)	total: 3m 18s	remaining: 47.4s
4500:	learn: 0.7360824	test: 10.7567304	best: 10.7567304 (4500)	total: 3m 44s	remaining: 22.7s
4956:	learn: 0.6487775	test: 10.7205467	best: 10.7

In [189]:
test['Ammo_shots_pred_final'] = model.predict(test[X])

In [190]:
error(test['Ammo_shots'], test['Ammo_shots_pred_final'])

9.814242052469696
0.3562213003017183


Мы получили лучший результат  
В среднем кол-во патронов будет откланяться на **9.81**, это составляет **0.35%** отклонения нашей предсказанной переменной от настоящего значения